### PR TITLE
feat: expose training context to rubrics via request protocol

### PR DESCRIPTION
**Pull request by shriramc1**
## Summary

Adds an optional `training_context` dict that orchestrators can pass to rubrics before scoring. This enables step-aware reward functions (curriculum learning, penalty warmup, dynamic weights) without requiring environments to maintain internal step counters or process restarts via env-args-scheduler.

## Motivation

There is currently no way for a rubric to know the current training step. This forces environment authors to use fragile workarounds like self-incrementing counters that don't survive checkpoint resume and don't reflect the true orchestrator step. Use cases blocked by this gap:

- **Reward warmup**: ramping penalty weights over training (e.g. tool call penalty warmup)
- **Curriculum learning**: changing reward thresholds or component weights based on progress
- **Adaptive scoring**: adjusting difficulty based on training metrics

The existing `env_args_scheduler` (PR #2207 in prime-rl) solves this by hot-reloading entire environments, which is too heavy for continuous reward parameter changes.

## Design

The `training_context` is a simple `dict | None` that flows through both execution paths:

**Server mode (ZMQ):**
```
orchestrator → EnvClient.run_group(training_context={"step": N})
  → RunGroupRequest(training_context={"step": N})
    → env_worker sets rubric.training_context before scoring
```

**Local mode (in-process):**
```
Environment.run_group(training_context={"step": N})
  → sets self.rubric.training_context before scoring
```

**Usage in a custom rubric:**
```python
class MyRubric(vf.Rubric):
    async def score_group(self, states):
        step = (self.training_context or {}).get("step", 0)
        warmup_progress = min(1.0, step / 100)
        penalty = -0.05 * warmup_progress
        # ... use penalty in scoring
```

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test improvement

## Testing

- [x] All existing tests pass when running `uv run pytest` locally.
- [ ] New tests have been added to cover the changes

## Checklist
- [x] My code follows the style guidelines of this project as outlined in [AGENTS.md](https://github.com/PrimeIntellect-ai/verifiers/blob/main/AGENTS.md)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published

## Changes

| File | Change |
|------|--------|
| `verifiers/serve/types.py` | Add optional `training_context: dict \| None = None` to `RunRolloutRequest` and `RunGroupRequest` |
| `verifiers/rubrics/rubric.py` | Add `self.training_context: dict \| None = None` to `Rubric.__init__` |
| `verifiers/serve/client/env_client.py` | Accept and forward `training_context` in `run_rollout` and `run_group` |
| `verifiers/serve/server/env_worker.py` | Set `self.env.rubric.training_context` before handling requests |
| `verifiers/envs/environment.py` | Accept `training_context` in `run_rollout`/`run_group`, set on rubric in local mode, forward to env_client in server mode |

## Backward Compatibility

Fully backward-compatible:
- All new parameters default to `None`
- Existing orchestrators that don't send `training_context` see zero behavior change
- Existing rubrics that don't read `self.training_context` are unaffected
- Wire protocol: old servers receiving a request with the extra field will ignore it (Pydantic `model_config` is not strict); old clients sending requests without it work fine since the field has a default

## Companion PR

The prime-rl companion PR (to populate `training_context` from the scheduler) is independent — this PR is useful standalone for:
- Local-mode environments (set `training_context` directly in `run_group` calls)
- Custom orchestrators built on verifiers
- Any user who wants step-aware rubrics without prime-rl

## Additional Notes

See also: prime-rl companion PR that populates this field from the scheduler's step counter.

---

> [!NOTE]
> **Medium Risk**
> Adds new mutable `training_context` plumbing through local and server execution paths; incorrect propagation or concurrent request handling could cause context leakage between rollouts/groups.
> 
> **Overview**
> Introduces an optional `training_context: dict | None` that can be passed into `Environment.run_rollout`/`run_group` and carried through the ZMQ request protocol.
> 
> In local mode, the context is set on `self.rubric.training_context` before scoring; in server mode it is forwarded via `RunRolloutRequest`/`RunGroupRequest` and applied in `EnvWorker` before executing the request. Test stubs were updated to accept extra kwargs for the extended call signatures.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bf0a711eeda91ddceb78a13d28af70524243f41d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: verifiers/serve/server/env_worker.py [hunk 1]
```diff
@@ -135,9 +135,9 @@ async def resolve_client(self, client_config: ClientConfig) -> Client:
             self.clients[key] = resolve_client(resolved)
         return self.clients[key]
 
-    async def handle_run_rollout(
-        self, request: RunRolloutRequest
-    ) -> RunRolloutResponse:
+    async def handle_run_rollout(self, request: RunRolloutRequest) -> RunRolloutResponse:
+        if request.training_context is not None:
+            self.env.rubric.training_context = request.training_context
```
[the remark this answers is not in the document]
Comment (shriramc1) [reply verifiers/serve/server/env_worker.py]: Addressed in the latest push (cd2700c2):

1. **Stale context**: Removed the `if not None` guard — we now always assign `training_context` (even when `None`), so each request explicitly sets or clears it.

2. **Concurrency**: In the current architecture, each `EnvWorker` processes requests sequentially through its event loop — `training_context` is set immediately before the scoring call within the same coroutine, so interleaving isn't possible within a single worker. The router distributes groups round-robin across workers, so no two concurrent requests share a rubric instance.

File: verifiers/envs/environment.py [hunk 2]
```diff
@@ -693,8 +674,12 @@ async def run_rollout(
                 sampling_args,
                 max_retries,
                 state_columns,
+                training_context=training_context,
             )
 
+        if training_context is not None:
+            self.rubric.training_context = training_context
```
[the remark this answers is not in the document]
Comment (shriramc1) [reply verifiers/envs/environment.py]: Addressed in the latest push (cd2700c2):

Added a property override in `RubricGroup` that propagates `training_context` to all child rubrics on assignment. This ensures user rubrics wrapped in a `RubricGroup` (which is the standard path for `MultiTurnEnv`, `ToolEnv`, etc.) receive the context correctly.

File: verifiers/rubrics/rubric.py [hunk 3]
```diff
@@ -48,6 +49,10 @@ def __init__(
 
         self.parser = parser or vf.Parser()
 
+        # Training context set by the orchestrator before scoring.
+        # Contains metadata like {"step": int, "ckpt_step": int}.
+        self.training_context: dict | None = None
```
[the remark this answers is not in the document]
Comment (shriramc1) [reply verifiers/rubrics/rubric.py:53]: Acknowledged — will add documentation in a follow-up once the API stabilizes through review. The feature is intentionally minimal right now (optional dict, defaults to None) so existing code is unaffected.
